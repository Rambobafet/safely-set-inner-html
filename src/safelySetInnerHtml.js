import React from 'react';
import {html2json} from 'html2json';

/**
 * Default config
 * @type {{ALLOWED_TAGS: string[]}}
 */
export const config = {
  ALLOWED_TAGS: [
    'strong',
    'a'
  ]
};

let tagId = 0;

/**
 * Based on html2json api, recursive function used to generate the react dom
 * @param node - node type (can be : root, element, text)
 * @param child - if the current node has children, it will be a filled array
 * @param text - if the current node has text
 * @param tag - if the current node is a tag
 * @param attr - if the current node has html attributes
 * @param allowedTags - map of authorized tags
 * @returns {Object}
 */
const generateDom = ({
  node = '',
  child = [],
  text = '',
  tag = '',
  attr = {}
}) => {
  const children = child.length ? child.map(generateDom) : text;

  if (node === 'element' && config.ALLOWED_TAGS.includes(tag)) {
    return React.createElement(tag, { ...attr, key: `safely-tag-${tagId++}` }, children);
  }

  return children;
};

/**
 * Basic API, will transform the given string in React DOM
 * @param str - html string to transform
 * @param allowedTags - custom authorized tags, not mandatory
 * @returns {Object}
 */
const safelySetInnerHtml = (str) => generateDom(html2json(str));

export default safelySetInnerHtml;
