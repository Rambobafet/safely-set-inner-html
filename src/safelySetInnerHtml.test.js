import React from 'react';
import renderer from 'react-test-renderer';
import safelySetInnerHtml, { config } from './safelySetInnerHtml';

const defaultConfig = { ...config };

describe('safelySetInnerHtml', () => {
  beforeEach(() => {
    Object.assign(config, defaultConfig);
  });

  it('renders correctly', () => {
    const dom = safelySetInnerHtml('Hello <strong>World !</strong>');
    const tree = renderer.create(<p>{dom}</p>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders only allowedTags', () => {
    config.ALLOWED_TAGS = [];

    const dom = safelySetInnerHtml('Hello <strong>World !</strong>');
    const tree = renderer.create(<p>{dom}</p>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const dom = safelySetInnerHtml('About the <a href="http://example.com">Author !</a>');
    const tree = renderer.create(<cite>{dom}</cite>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
