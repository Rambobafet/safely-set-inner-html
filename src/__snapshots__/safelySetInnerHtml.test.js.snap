// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`safelySetInnerHtml renders correctly 1`] = `
<p>
  Hello 
  <strong>
    World !
  </strong>
</p>
`;

exports[`safelySetInnerHtml renders correctly 2`] = `
<cite>
  About the 
  <a
    href="http://example.com"
  >
    Author !
  </a>
</cite>
`;

exports[`safelySetInnerHtml renders only allowedTags 1`] = `
<p>
  Hello 
  World !
</p>
`;
